src/cfg/cfg-traversal.h

```diff
@@ -296,11 +296,13 @@ struct CFGWalker : public ControlFlowWalker<SubType, VisitorType> {
 
   static void doEndCall(SubType* self, Expression** currp) {
     doEndThrowingInst(self, currp);
-    if (!self->throwingInstsStack.empty()) {
-      // exception not thrown. link to the continuation BB
-      auto* last = self->currBasicBlock;
-      self->link(last, self->startBasicBlock());
-    }
+    // Create a new basic block and link to it. We do this even if there are no
+    // other edges leaving this call (no catch bodies in this function that we
+    // can reach if we throw), because we want to preserve the property that a
+    // basic block ends with an instruction that might branch, and the call
+    // might branch out of the entire function if it throws.
+    auto* last = self->currBasicBlock;
+    self->link(last, self->startBasicBlock());
```

Comment:
Hmm, I think I decided not to do this because it would create too many edges. What changed, and what's the next PR this helps?

And maybe we can at least check if EH is available?

Reply:
Good point about checking for EH, that should help... I should also measure the overhead.

I think what changed is the new pass I am working on :smile: but I think it's a general principle that makes sense, that control flow can't move in the middle of a block (except for traps). LLVM etc. do that, I think?

The new pass is an analysis for TrapsNeverHappen that goes backwards. Imagine a function always casts the input parameter to type B, even though it is declared as a supertype A. That is common in Java etc., where class methods declare `this` to have a very generic type like `java.lang.Object`, while methods of a class `B` will immediately cast to that type. If we see a static call to such a function, then we can infer in the caller that the reference must have that type (if we assume traps never happen).

And that pass needs to take into account transfers of control flow:
```wat
(call $will-cast-to-B
  (block
    (local.get $ref) ;; we can't infer this far back
    (call $maybe-throw)
    (local.get $ref) ;; we can infer about this
  )
)

(func $will-cast-to-B (param $ref anyref)
  (ref.cast $B (local.get $ref))
```
Imagine that `maybe-throw` will throw if that local doesn't have the right type `B`. If so then we won't actually do the call, which means we can't infer the first get has type `B` (only the second). That is, we can only make this inference when the call will definitely execute from that point.

And I am checking if the call happens by looking just at the same basic block as it. That is, when I see a parameter that is in another basic block, I stop going backwards, as control flow might go elsewhere. For that, I need to know that a call can transfer control flow.

Reply:
Ah, makes sense. Thanks for explaining.